### ISSUE TITLE
[question] Editable packages and lockfiles - How do they work together?

### ISSUE THREAD

**Issue by opajonk** [conan-io/conan#15942]
### What is your question?

I'm trying to get a build with an editable dependency to work together with a lockfile. The lockfile contains several dependencies, building with it works. Now, one of the dependencies should be made into an editable package. Building with `conan build . --build=editable` works, it uses the editable package, but it also takes random other dependencies as no lockfile is given. On the other hand, running `conan build . --build=editable --lockfile=mydeps.lock` seems to ignore the `--build=editable` part and just uses the lockfile. I can work around it by removing the relevant entry from the lockfile (`conan lock remove --requires=libA/* --lockfile=mydeps.lock --lockfile-out=myreduceddeps.lock` and then running the build with `conan build . --build=editable --lcckfile=myreduceddeps.lock --lockfile-partial`, but I have a hard time to tell the devs on my team that this is the way to do it.

That editable packages do not "override" lockfiles seems to be strage - other package managers behave here differently; they use the "local dependency" in any case, if one is given.
 
What is the intended workflow to get locally editable dependencies to work together with dependencies defined via lockfile?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @opajonk 

Thanks for your report.

> On the other hand, running conan build . --build=editable --lockfile=mydeps.lock seems to ignore the --build=editable part and just uses the lockfile.

Do you mean that in that case, the dependency is used from the Conan cache, instead of the editable one?

If that is the case, this sounds something to have a look and try to improve, indeed.

**Comment by opajonk** [comment]
I just double-checked with my colleague, and yes this seems to be the case - it uses the one from the cache, as given in the lockfile. Not the "editable" one...

**Comment by memsharded** [comment]
I am trying to reproduce, this test seems to be working fine so far:

```python
def test_editable_lockfile():
    """
    https://github.com/conan-io/conan/issues/15942
    """
    c = TestClient()
    c.save({"pkga/conanfile.py": GenConanfile("pkga", "1.0"),
            "app/conanfile.py": GenConanfile("app", "1.0").with_requires("pkga/1.0")})
    c.run("create pkga")
    c.run("install app --lockfile-out=conan.lock")

    # now we want to investigate pkga
    c.run("editable add pkga")
    pkga = textwrap.dedent("""
        from conan import ConanFile
        class PkgA(ConanFile):
            name = "pkga"
            version = "1.0"
            def build(self):
                self.output.info("BUILDING PKGA!!!")
            def package_info(self):
                self.output.info("INFO PKGA!!!")
        """)
    c.save({"pkga/conanfile.py": pkga})
    
    c.run("install app --lockfile=conan.lock")
    assert "pkga/1.0: BUILDING PKGA!!!" not in c.out
    assert "pkga/1.0: INFO PKGA!!!" in c.out
    
    c.run("install app --build=editable --lockfile=conan.lock")
    assert "Using lockfile" in c.out
    assert "pkga/1.0: BUILDING PKGA!!!" in c.out
    assert "pkga/1.0: INFO PKGA!!!" in c.out
```

Can you please check it? I hope it is not difficult to understand. It would be good to have detailed instructions how to reproduce what you are seeing.

**Comment by opajonk** [comment]
Hmm maybe an explanation can be that we use a `global.conf` with

```
core.package_id:default_embed_mode = full_mode
core.package_id:default_non_embed_mode = full_mode
core.package_id:default_unknown_mode = full_mode
```

?

We have to do that in order to be able to work with branches over multiple repositories in parallel.

**Comment by memsharded** [comment]
Not apparently, it also works with that:
```python
def test_editable_lockfile():
    """
    https://github.com/conan-io/conan/issues/15942
    """
    c = TestClient()
    global_conf = textwrap.dedent("""\
        core.package_id:default_embed_mode = full_mode
        core.package_id:default_non_embed_mode = full_mode
        core.package_id:default_unknown_mode = full_mode
        """)
    c.save_home({"global.conf": global_conf})
    c.save({"pkga/conanfile.py": GenConanfile("pkga", "1.0"),
            "app/conanfile.py": GenConanfile("app", "1.0").with_requires("pkga/1.0")})
    c.run("create pkga")
    c.run("install app --lockfile-out=conan.lock")

    # now we want to investigate pkga
    c.run("editable add pkga")
    pkga = textwrap.dedent("""
        from conan import ConanFile
        class PkgA(ConanFile):
            name = "pkga"
            version = "1.0"
            def build(self):
                self.output.info("BUILDING PKGA!!!")
            def package_info(self):
                self.output.info("INFO PKGA!!!")
        """)
    c.save({"pkga/conanfile.py": pkga})

    c.run("install app --lockfile=conan.lock")
    assert "pkga/1.0: BUILDING PKGA!!!" not in c.out
    assert "pkga/1.0: INFO PKGA!!!" in c.out

    c.run("install app --build=editable --lockfile=conan.lock")
    assert "Using lockfile" in c.out
    assert "pkga/1.0: BUILDING PKGA!!!" in c.out
    assert "pkga/1.0: INFO PKGA!!!" in c.out
```

**Comment by memsharded** [comment]
As a side note, I guess that you are aware that ``core.package_id:default_non_embed_mode = full_mode`` is quite aggressive (the other two are fine, the first one is actually the default). If you have enough CI power and it is not an issue for you, fine, but this shouldn't be necessary for many cases and the time and cost savings can be very relevant.

**Comment by opajonk** [comment]
Hmm interesting. We will investigate now what goes wrong in our case. Thanks a lot again for the quick response - I will come back when I have an update!

**Comment by opajonk** [comment]
So, we did investigate and found an "interesting" issue. In our case, `pkga` had an overridden `set_version()`, which added build metadata to the version string (say, `1.0.0+<git-sha>`). This is included in the lockfile, of course. It is also included in the editable package reference. However, if you now create a git commit in `pkga`, it gets a new `<git-sha>` (of course), and now - depending on the order of commands, you end up setting a package to editable that has a different version string than the one referenced in the lockfile of `app`. And that does not work; the build of `app` uses the one from the cache. I hope this was not too confusing ;-)

The question is: should build-metadata be part of the "editable package reference"? I am not sure what makes most sense, to be honest...

**Comment by memsharded** [comment]
I am afraid that I can't still reproduce.
I think the following test captures the case that you are describing, yet passes fine:

```python
def test_editable_lockfile():
    """
    https://github.com/conan-io/conan/issues/15942
    """
    c = TestClient()
    global_conf = textwrap.dedent("""\
        core.package_id:default_embed_mode = full_mode
        core.package_id:default_non_embed_mode = full_mode
        core.package_id:default_unknown_mode = full_mode
        """)
    c.save_home({"global.conf": global_conf})
    c.save({"pkga/conanfile.py": GenConanfile("pkga", "1.0+b.1"),
            "app/conanfile.py": GenConanfile("app", "1.0").with_requires("pkga/[>=1.0]")})
    c.run("create pkga")
    c.run("install app --lockfile-out=conan.lock")
    print(c.load("conan.lock"))

    # now we want to investigate pkga
    c.run("editable add pkga")
    pkga = textwrap.dedent("""
        from conan import ConanFile
        class PkgA(ConanFile):
            name = "pkga"
            version = "1.0+b.2"

            def build(self):
                self.output.info("BUILDING PKGA!!!")
            def package_info(self):
                self.output.info("INFO PKGA!!!")
        """)
    c.save({"pkga/conanfile.py": pkga})

    c.run("install app --lockfile=conan.lock")
    print(c.out)
    assert "pkga/1.0+b.1: BUILDING PKGA!!!" not in c.out
    assert "pkga/1.0+b.1: INFO PKGA!!!" in c.out

    c.run("install app --build=editable --lockfile=conan.lock")
    print(c.out)
    assert "Using lockfile" in c.out
    assert "pkga/1.0+b.1: BUILDING PKGA!!!" in c.out
    assert "pkga/1.0+b.1: INFO PKGA!!!" in c.out
```

In any case I can make it go to the Conan cache to resolve the package instead of the editable one.

**Comment by memsharded** [comment]
Hi @opajonk 

Any follow-up here? Did you manage to see my updated test above? Thanks!

**Comment by opajonk** [comment]
Hi! Sorry for the (very) late reply here.

Yes, we checked the tests, and I think in the end you were right. I am not sure about the build metadata anymore, and unfortunately I cannot reproduce this now :-(

Anyway, thanks a lot for your help and timely replies (in contrast to me) - this is much appreciated!

**Comment by memsharded** [comment]
> Hi! Sorry for the (very) late reply here.

Dont worry about that! 🙂 I know we are all very busy, we also sometimes fail to follow up on tickets, don't hesitate to ping us if ever necessary because we don't respond on time, to resurface the ticket, it happens!

> Yes, we checked the tests, and I think in the end you were right. I am not sure about the build metadata anymore, and unfortunately I cannot reproduce this now :-(

No problem, just re-open or create a new ticket if you happen to have a similar issue and we will take it from there again.

> Anyway, thanks a lot for your help and timely replies (in contrast to me) - this is much appreciated!

Thanks to you for the feedback!
